src/App.tsx

```diff
@@ -15,48 +16,120 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  NoSort,
+  Name,
+  Length,
+}
+
+enum DisplayOrder {
+  Normal,
+  Reverse,
+}
+
+type DisplayOptions = {
+  sortBy: SortType;
+  displayOrder: DisplayOrder;
+};
+
+function getPreparedGoods(
+  goods: string[],
+  { sortBy, displayOrder }: DisplayOptions,
+) {
+  const preparedGoods = [...goods];
+
+  if (sortBy !== SortType.NoSort) {
+    preparedGoods.sort((a, b) => {
+      switch (sortBy) {
+        case SortType.Name:
+          return a.localeCompare(b);
+
+        case SortType.Length:
+          return a.length - b.length;
+
+        default:
+          return 0;
+      }
+    });
+  }
+
+  if (displayOrder === DisplayOrder.Reverse) {
+    preparedGoods.reverse();
+  }
+
+  return preparedGoods;
+}
+
 export const App: React.FC = () => {
+  const [sortBy, setSortBy] = useState(SortType.NoSort);
+  const [displayOrder, setDisplayOrder] = useState(DisplayOrder.Normal);
+
+  const toggleDisplayOrder = () => {
+    setDisplayOrder(
+      displayOrder === DisplayOrder.Normal
+        ? DisplayOrder.Reverse
+        : DisplayOrder.Normal,
+    );
```

Comment:
In order to always work correctly with the current state, it is necessary to take it from the callback argument.
```suggestion
    setDisplayOrder(prevOrder =>
      prevOrder === DisplayOrder.Normal
        ? DisplayOrder.Reverse
        : DisplayOrder.Normal,
    );
```